compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -67,9 +68,21 @@ import Development.IDE.Types.Location
 -- So we have our own wrapper here that supports empty filepaths.
 uriToFilePath' :: Uri -> Maybe FilePath
 uriToFilePath' uri
-    | uri == filePathToUri "" = Just ""
+    | uri == filePathToUri' "" = Just ""
     | otherwise = LSP.uriToFilePath uri
 
+-- TODO This is a temporary hack: VSCode escapes ':' in URIs while haskell-lsp’s filePathToUri doesn't.
+-- This causes issues since haskell-lsp stores the original URI in the VFS while we roundtrip once via
+-- uriToFilePath' and filePathToUri before we look it up again. At that point : will be unescaped in the URI
+-- so the lookup fails. The long-term solution here is to avoid roundtripping URIs but that is a larger task
+-- so for now we have our own version of filePathToUri that does escape colons.
+filePathToUri' :: FilePath -> Uri
+filePathToUri' fp =
+    case T.stripPrefix "file:" (getUri uri) of
+        Just suffix -> Uri $ T.pack $ "file:" <> escapeURIString (/= ':') (T.unpack suffix)
+        Nothing -> uri
+    where uri = filePathToUri fp
```

Comment:
I guess `filePathToUri` has some LSP-specific logic, otherwise I would recommend using `escapeURIString isUnescapedInURIComponent` instead of building the URI / stripping the prefix / escaping / readding prefix.

Reply:
Not really LSP specific but it takes care of converting backslashes to forward slashes which is necessary for `file:` URIs.

Reply:
Right! Filepaths... !!